realm/realm-library/src/objectServer/java/io/realm/Progress.java

```diff
@@ -0,0 +1,131 @@
+ * {@link #getFractionTransferred()} can both increase and decrease since more changes might be added while
+ * the progres listener is registered. This means it is possible for one notification to report
+ * {@code true} for {@link #isTransferComplete()}, and then on the next event report {@code false}.
+ * <p>
+ * If the {@link ProgressListener} was registered with {@link ProgressMode#CURRENT_CHANGES}, progress can only ever
+ * increase, and once {@link #isTransferComplete()} returns {@code true}, no further events will be generated.
+ *
+ * @see SyncSession#addDownloadProgressListener(ProgressMode, ProgressListener)
+ * @see SyncSession#addUploadProgressListener(ProgressMode, ProgressListener)
+ */
+public class Progress {
+
+    private final long transferredBytes;
+    private final long transferableBytes;
+
+    /**
+     * Creates a snapshot of the current progress when downloading or uploading changes.
+     *
+     * @param transferredBytes number of bytes transferred.
+     * @param transferableBytes total number of bytes that needs to be transferred (including those already transferred).
+     */
+    Progress(long transferredBytes, long transferableBytes) {
+        this.transferredBytes = transferredBytes;
+        this.transferableBytes = transferableBytes;
+    }
+
+    /**
+     * Returns the total number of bytes that has been transferred since the {@link ProgressListener} was added.
+     *
+     * @return the total number of bytes transferred since the {@link ProgressListener} was added.
+     */
+    public long getTransferredBytes() {
+        return transferredBytes;
+    }
+
+    /**
+     * Returns the total number of transferable bytes (bytes that have been transferred + bytes pending transfer).
+     * <p>
+     * If the {@link ProgressListener} is tracking downloads, this number represents the size of the changesets
+     * generated by all other clients using the Realm.
+     * <p>
+     * If the {@link ProgressListener} is tracking uploads, this number represents the size of changesets created
+     * locally.
+     *
+     * @return the total number of bytes that has been transferred + number of bytes still pending transfer.
+     */
+    public long getTransferableBytes() {
+        return transferableBytes;
+    }
+
+    /**
+     * The fraction of bytes transferred out of all transferable bytes. Counting from since the {@link ProgressListener}
+     * was added.
+     *
+     * @return a number between {@code 0.0} and {@code 1.0}, where {@code 0.0} represents that no data has been
+     *         transferred yet, and {@code 1.0} that all data has been transferred.
+     */
+    public double getFractionTransferred() {
+        if (transferableBytes == 0) {
+            return 1.0D;
+        } else {
+            double percentage = (double) transferredBytes / (double) transferableBytes;
+            return percentage > 1.0D ? 1.0D : percentage;
```

Comment:
Mostly to be extra safe and there is no restriction to `transferredBytes` / `transferableBytes`. If we lift this, we should probably guard the constructor?